package/network/config/ltq-vdsl-app/files/10_atm_ptm.sh

```diff
@@ -0,0 +1,51 @@
+#!/bin/sh
+
+[ "$DSL_NOTIFICATION_TYPE" = "DSL_STATUS" ] || exit 0
+
+if [ "$DSL_TC_LAYER_STATUS" = "ATM" ]; then
+
+	if grep -q "ltq_ptm_vr9" /proc/modules ; then
+		# TODO: make ltq_atm_vr9/ltq_ptm_vr9 "hot-swappable"
+		# ifconfig ptm0 down &>/dev/null
+		# rmmod "ltq_ptm_vr9"
+		logger -p daemon.notice -t "dsl-notify" "Switching to TC-Layer ATM while PTM driver is loaded. Rebooting System"
+		reboot
+		exit
+	fi
+
+	if ! grep -q "ltq_atm_vr9" /proc/modules ; then
+		logger -p daemon.notice -t "dsl-notify" "Switching to TC-Layer ATM"
+
+		modprobe ltq_atm_vr9
+		esi $(printf '%012X' $((1+0x$(tr -d : </sys/class/net/eth0/address))))
+		
+		grep -q "br2684" /proc/modules || modprobe br2684
+		/etc/init.d/br2684ctl start &>/dev/null
+		ip link set dev nas0 up
+	fi
+
+elif [ "$DSL_TC_LAYER_STATUS" = "EFM" ]; then
+	if grep -q "br2684" /proc/modules ; then
+		ifconfig nas0 down &>/dev/null
+		/etc/init.d/br2684ctl stop &>/dev/null
+		rmmod "br2684"
+	fi
+
+	if grep -q "ltq_atm_vr9" /proc/modules ; then
+		# TODO: make ltq_atm_vr9/ltq_ptm_vr9 "hot-swappable"
+		# rmmod "ltq_atm_vr9"
+		logger -p daemon.notice -t "dsl-notify" "Switching to TC-Layer EFM/PTM while ATM driver is loaded. Rebooting System"
+		reboot
+		exit
+	fi
+
+	if ! grep -q "ltq_ptm_vr9" /proc/modules ; then
+		logger -p daemon.notice -t "dsl-notify" "Switching to TC-Layer EFM/PTM"
+
+		modprobe ltq_ptm_vr9
+		ip link set dev ptm0 down
+		ip link set dev ptm0 address $(printf '%012X' $((1+0x$(tr -d : </sys/class/net/eth0/address))))
```

Comment:
> In my current situation I don't want to use the ucidef_set_interface_macaddr() because...

Well, ucidef_set_interface_macaddr() is heavily used in the lantiq target. Even if you're  not intending to use the function, you would break exists code/configs. Please test your commits for possible breakage.

> Doing this with the mac in /e/c/network will result in having a lot of devices with the same macs

I'm well aware and already hit by the same issue. We need something that applies default settings without adding them to the config files. It is a treewide problem and makes it for example unnecessary hard to rename leds. So far we are aware of the limitation but no one has started to work on this.

> I've also integrated this based on the esi command in the atm section witch is taken from the vdsl_cpe_control_wrapper. Isn't this also faulty then?

No it isn't. As far as I know, the esi command was added because of a race condition. In a dhcp over atm configuration the nas0 interface was up before netifd applied the mac address. With the result that some bogus mac address was used in the dhcp packages. The "fix" is more or less an ugly hack. Especially since the ucidef_set_interface_macaddr() mac address is ignored for atm bridges.

To my knowledge there is no real atm bridge integration in netifd. That is why br2684ctl is used. With all the drawbacks like dealing with race conditions (f6e38ec125ddb1da51780e510c7943e76f61f88d).

It might be than the linked commit already fixed the dhcp over atm race condition. Unfortunally the user with the issue isn't available for testing/doesn't have the dhcp over atm line anymore.